src/net/protocol_agnostic/listener.rs

```diff
@@ -140,37 +191,132 @@ impl PaListener {
     pub fn local_addr(&self) -> io::Result<PaAddr> {
         match *self {
             PaListener::Tcp(ref tcp_listener) => Ok(PaAddr::Tcp(tcp_listener.local_addr()?)),
-            PaListener::Utp(ref utp_listener) => Ok(PaAddr::Utp(utp_listener.local_addr()?)),
+            PaListener::Utp(_, ref utp_listener) => Ok(PaAddr::Utp(utp_listener.local_addr()?)),
         }
     }
 
     pub fn incoming(self) -> PaIncoming {
         match self {
-            PaListener::Tcp(tcp_listener) => PaIncoming::Tcp(tcp_listener.incoming()),
-            PaListener::Utp(utp_listener) => PaIncoming::Utp(utp_listener.incoming()),
+            PaListener::Tcp(tcp_listener) => {
+                let incoming = tcp_listener.incoming();
+                let processing = FuturesUnordered::new();
+                PaIncoming {
+                    inner: PaIncomingInner::Tcp {
+                        incoming,
+                        processing,
+                    },
+                }
+            },
+            PaListener::Utp(socket, utp_listener) => {
+                let incoming = utp_listener.incoming();
+                let raw_rx = socket.raw_receiver();
+                let processing = FuturesUnordered::new();
+                PaIncoming {
+                    inner: PaIncomingInner::Utp {
+                        incoming,
+                        raw_rx,
+                        processing,
+                    },
+                }
+            },
         }
     }
 }
 
 impl Stream for PaIncoming {
     type Item = (PaStream, PaAddr);
-    type Error = io::Error;
+    type Error = AcceptError;
 
-    fn poll(&mut self) -> io::Result<Async<Option<(PaStream, PaAddr)>>> {
-        match *self {
-            PaIncoming::Tcp(ref mut incoming) => {
-                match incoming.poll()? {
-                    Async::Ready(Some((stream, addr))) => {
-                        let stream = PaStream::Tcp(stream);
-                        let addr = PaAddr::Tcp(addr);
-                        Ok(Async::Ready(Some((stream, addr))))
+    fn poll(&mut self) -> Result<Async<Option<(PaStream, PaAddr)>>, AcceptError> {
+        match self.inner {
+            PaIncomingInner::Tcp { ref mut incoming, ref mut processing } => {
+                loop {
+                    match incoming.poll().map_err(AcceptError::TcpAccept)? {
+                        Async::Ready(Some((stream, addr))) => {
+                            processing.push({
+                                let header = [0u8; 8];
```

Comment:
This body of this match arm could be extracted to separate function. That would make code a little more readable, maintainable, testable, .etc